### PR TITLE
WiFi.config handle Arduino parameters ordering and auto dns,gw,mask

### PR DESCRIPTION
**Pull request by JAndrassy**
repeat of https://github.com/espressif/arduino-esp32/pull/8892 rebased after networking refactoring.

In WiFi libraries by Arduino `config` has parameters `ip, dnsIP, gatewayIP, subnet`. only ip is not optional. for unspecified subnet default is 255.255.255.0 and unspecified gateway and dns are based on IP with last number changed to 1.

the ESP8266WiFi library accepts the 'esp\ ordering ip, gw, subnet, dns and Arduino ordering ip, dns, gw, subnet. The detection is based on the position of the subnet mask which has 255 as first number.

This PR adds handling of Arduino ordering of parameters. Parameter subnet is on position of dns1 which is an optional parameter so default value is used if needed.

And this PR adds a second `config` with two parameters local_ip and dns IP. dns IP is optional. gateway IP is calculated, dns IP is calculated if not provided. then the other `config` is invoked.

this allows `config` invocations as in Arduino library
```
config(ip)
config(ip, dns)
config(ip, dns, gw)
config(ip, dns, gw, subnet)
```
it also allows `WiFi.config(INADDR_NONE)` to return to DHCP as in other WiFi libraries.

overview of WiFi object API in significant libraries: 
https://github.com/JAndrassy/Arduino-Networking-API/blob/main/ArduinoNetAPILibs.md#wifi-station-network-interface

### CODE REVIEW

File: libraries/WiFi/src/WiFiSTA.cpp [hunk 1]
```diff
@@ -169,9 +169,33 @@ bool WiFiSTAClass::eraseAP(void) {
...
+        gateway = subnet;
+        subnet = (tmp != INADDR_NONE) ? tmp : IPAddress(255, 255, 255, 0);
+    }
+
     return STA.config(local_ip, gateway, subnet, dns1, dns2);
 }
 
+bool WiFiSTAClass::config(IPAddress local_ip, IPAddress dns) {
+
+    if (local_ip == INADDR_NONE)
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiSTA.cpp:185]: please do not use short `if`

File: libraries/WiFi/src/WiFiSTA.cpp [hunk 2]
```diff
@@ -169,9 +169,33 @@ bool WiFiSTAClass::eraseAP(void) {
...
+        return STA.config(INADDR_NONE, INADDR_NONE, INADDR_NONE);
+
+    if (local_ip.type() != IPv4)
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiSTA.cpp:188]: please do not use short `if`

File: libraries/WiFi/src/WiFiSTA.cpp [hunk 3]
```diff
@@ -169,9 +169,33 @@ bool WiFiSTAClass::eraseAP(void) {
...
+        subnet = (tmp != INADDR_NONE) ? tmp : IPAddress(255, 255, 255, 0);
+    }
+
     return STA.config(local_ip, gateway, subnet, dns1, dns2);
 }
 
+bool WiFiSTAClass::config(IPAddress local_ip, IPAddress dns) {
+
+    if (local_ip == INADDR_NONE)
+        return STA.config(INADDR_NONE, INADDR_NONE, INADDR_NONE);
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiSTA.cpp:186]: Return `config(INADDR_NONE, INADDR_NONE, INADDR_NONE);` instead

File: libraries/WiFi/src/WiFiSTA.cpp [hunk 4]
```diff
@@ -169,9 +169,33 @@ bool WiFiSTAClass::eraseAP(void) {
...
+
+    if (local_ip.type() != IPv4)
+        return false;
+
+    IPAddress gw(local_ip);
+    gw[3] = 1;
+    if (dns == INADDR_NONE) {
+        dns = gw;
+    }
+    return STA.config(local_ip, gw, IPAddress(255, 255, 255, 0), dns);
```
Comment (me-no-dev) [inline libraries/WiFi/src/WiFiSTA.cpp:196]: Return `STA.config(local_ip, gw, IPAddress(255, 255, 255, 0), dns);` instead

Comment (JAndrassy) [reply libraries/WiFi/src/WiFiSTA.cpp:196]: @me-no-dev please clarify this one

**Pull request merged**